### PR TITLE
Identify language using a separate AI prompt

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part of: #10103

## Description

Tries to identify the language using a separate prompt. Inputs the identified language to the text generation prompt.

### Notes
I think we should skip storing the identified language due to the following reasons.
1. A store might have products in different languages.
2. If incorrect language is identified for a product the first time, we will keep generating text in the wrong language for all other products. (which will be a pretty bad user experience)

Should we cache the language information to reduce network requests?

## Testing instructions

- Log in to a published WPCom site.
- Select an existing product or create a new one.
- Select Write with AI in the description row.
- On the description AI sheet, select Write with AI. The response should be in the same language as the product title and features.
- Tap Regenerate and make sure that new responses are still good.
- Tap the `...` button from the top right and tap Share to open the share sheet
- Tap "Write with AI" to generate a "share message".
- The generated share message should be in the same language as the product title and description.

## Screenshots
| Product description | Product sharing |
|--------|--------|
| Simulator Screen Shot - iPhone 14 Pro - 2023-07-05 at 10 28 26 | Simulator Screen Shot - iPhone 14 Pro - 2023-07-05 at 10 28 15 |

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by itsmeichigo** [comment]
Thanks @selanthiraiyan for the idea and for demonstrating the solution! The result works well for all my products so far, for both product description and sharing! 💯 

I'm concerned about hitting the limit for AI requests though. I agree that we should not cache the identified language for the whole store, but how about this:
- Move the language identification into a separate action.
- Call it only once every time the main CTA is tapped and use the same language when the user taps the Regenerate button. This should work since product information stay fixed for product sharing; and even when users update product features for generating description, they should not change the product language entirely.

**Comment by selanthiraiyan** [comment]
Thanks for the review and testing, @itsmeichigo!

Good point about reusing the language for the "Regenerate" button. 👍 I have made the necessary changes to save the identified language. 

Please take another look at the PR. 🙇

File: Yosemite/Yosemite/Stores/ProductStore.swift [hunk 1]
```diff
@@ -531,20 +535,41 @@ private extension ProductStore {
...
+        Task {
+            let result = await Result {
+                try await generativeContentRemote.identifyLanguage(siteID: siteID,
+                                                                   string: string,
+                                                                   feature: feature)
+            }
+            await MainActor.run {
+                completion(result)
+            }
+        }
```
Comment (itsmeichigo) [inline Yosemite/Yosemite/Stores/ProductStore.swift]: Nit: I thought `await MainActor.run` is equivalent as setting `@MainActor` for the `Task`. Can we use it to simplify the code?

```suggestion
        Task { @MainActor in
            let result = await Result {
                try await generativeContentRemote.identifyLanguage(siteID: siteID,
                                                                   string: string,
                                                                   feature: feature)
            }
            completion(result)
        }
```

Comment (itsmeichigo) [reply Yosemite/Yosemite/Stores/ProductStore.swift]: If this makes sense, it's probably worth it to apply the same thing to the tasks  below.

File: Yosemite/Yosemite/Stores/ProductStore.swift [hunk 2]
```diff
@@ -531,20 +535,41 @@ private extension ProductStore {
...
+                                    language: String,
                                     completion: @escaping (Result<String, Error>) -> Void) {
-        let prompt = [
-            "Write a description for a product with title ```\(name)``` and features: ```\(features)```.",
-            "Identify the language used in the product title and features and use the same language in your response.",
-            "Make the description 50-60 words or less.",
-            "Use a 9th grade reading level.",
-            "Perform in-depth keyword research relating to the product in the same language of the product title, " +
-            "and use them in your sentences without listing them out."
-        ].joined(separator: "\n")
```
Comment (itsmeichigo) [inline Yosemite/Yosemite/Stores/ProductStore.swift]: ❓ Just curious: is there a reason why we should move the prompt inside of `Task`?

**Review by itsmeichigo — APPROVED** [review]
This works as expected! 🚢

File: Yosemite/Yosemite/Stores/ProductStore.swift (hunk 1 above)
Comment (selanthiraiyan) [reply Yosemite/Yosemite/Stores/ProductStore.swift]: I think the existing code structure doesn't force the `generativeContentRemote` to work on the main thread. 
Once we have the response, we send that back in the main thread using `await MainActor.run {`

Kindly let me know if I am missing something here.

File: Yosemite/Yosemite/Stores/ProductStore.swift (hunk 2 above)
Comment (selanthiraiyan) [reply Yosemite/Yosemite/Stores/ProductStore.swift]: I changed this when we were sending the request for language identification from this task. 

I moved the prompts outside in https://github.com/woocommerce/woocommerce-ios/pull/10169/commits/3154d66f96f49217bedf30e5e3382c52dc3143a1

File: Yosemite/Yosemite/Stores/ProductStore.swift (hunk 1 above)
Comment (itsmeichigo) [reply Yosemite/Yosemite/Stores/ProductStore.swift]: From my understanding, `generativeContentRemote` isn't forced to run on the main thread when we set `@MainActor` to the task. It's called with `await`, which is a suspension point, and from here, the code can be dispatched to any thread the device finds suitable. Adding `@MainActor` ensures that the result of the async code is dispatched to the main thread.

Comment (selanthiraiyan) [reply Yosemite/Yosemite/Stores/ProductStore.swift]: Thanks for explaining. Done in https://github.com/woocommerce/woocommerce-ios/pull/10169/commits/0c69711f36087b7fcbc1a24bb0dd34229e7c3785

**Pull request merged**